### PR TITLE
Beam Breaker -> core

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.utils.digitalinput.DigitalInputInputsAutoLogged;
+import frc.utils.digitalinput.IDigitalInput;
+
+public class ChooserDigitalInput implements IDigitalInput {
+
+	private SendableChooser<Boolean> sendableChooser;
+
+	public ChooserDigitalInput(String chooserName) {
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:12]:
```suggestion
	public ChooserDigitalInput(String name) {
```

File: src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		this.sendableChooser = new SendableChooser<>();
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:13]: line space after

File: src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		sendableChooser.setDefaultOption(
+			String.valueOf(ChooserDigitalInputConstants.DEFAULT_STATE),
+			ChooserDigitalInputConstants.DEFAULT_STATE
+		);
+		sendableChooser.addOption("true", true);
+		sendableChooser.addOption("false", false);
+
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:20]: remove line space

File: src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package frc.utils.digitalinput.chooser;
+
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.utils.digitalinput.DigitalInputInputsAutoLogged;
+import frc.utils.digitalinput.IDigitalInput;
+
+public class ChooserDigitalInput implements IDigitalInput {
+
+	private SendableChooser<Boolean> sendableChooser;
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:10]:
```suggestion
	private SendableChooser<Boolean> digitalInputValue;
```

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import frc.utils.digitalinput.IDigitalInput;
+
+import java.util.function.BooleanSupplier;
+
+public class SuppliedDigitalInput implements IDigitalInput {
+
+	private final Debouncer debouncer;
+	private boolean inverted;
+
+	private final BooleanSupplier booleanSupplier;
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:14]:
```suggestion
	private final Debouncer debouncer;
	private boolean inverted;

	private final BooleanSupplier digitalInputSupplier;
```

Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:14]:
```suggestion
	private final BooleanSupplier booleanSupplier;
	private final Debouncer debouncer;
	private boolean inverted;
```

File: src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+	public ChooserDigitalInput(String chooserName) {
+		this.sendableChooser = new SendableChooser<>();
+		sendableChooser.setDefaultOption(
+			String.valueOf(ChooserDigitalInputConstants.DEFAULT_STATE),
+			ChooserDigitalInputConstants.DEFAULT_STATE
+		);
+		sendableChooser.addOption("true", true);
+		sendableChooser.addOption("false", false);
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:19]: do you need  defalut and add? i think u need just to setDefualt and add the other one

Comment (Nitay4) [reply src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:19]: but the DEFULT_STATE is in the constants, I don't necessarily know which one to not add

File: src/main/java/frc/robot/Main.java [hunk 7]
```diff
@@ -19,6 +20,7 @@ public final class Main {
 	 */
 	public static void main(String... args) {
 		RobotBase.startRobot(RobotManager::new);
+		System.out.println(String.valueOf(ChooserDigitalInputConstants.DEFAULT_STATE));
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/Main.java:23]: really?

File: src/main/java/frc/robot/RobotManager.java [hunk 8]
```diff
@@ -54,6 +55,9 @@ public void robotPeriodic() {
 		CommandScheduler.getInstance().run();
 		BatteryUtils.logStatus();
 		BusChain.logChainsStatuses();
+		robot.channeledDigitalInput.updateInputs(robot.inputs);
+		Logger.recordOutput("isTrue", robot.inputs.debouncedValue);
+		Logger.recordOutput("nnnnnn", robot.inputs.nonDebouncedValue);
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/RobotManager.java:60]: remove debugging. This is a PR

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+	public SuppliedDigitalInput(
```
Comment (Sh-Pe) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:16]: I think you can overload this too

File: src/main/java/frc/utils/digitalinput/channeled/ChanneledDigitalInput.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import frc.utils.digitalinput.IDigitalInput;
+import org.littletonrobotics.junction.Logger;
+
+public class ChanneledDigitalInput implements IDigitalInput {
+
+	private final DigitalInput digitalInput;
+	private final Debouncer debouncer;
+	private final boolean inverted;
+
+	public ChanneledDigitalInput(int channel, double debounceTime, boolean inverted) {
```
Comment (Sh-Pe) [inline src/main/java/frc/utils/digitalinput/channeled/ChanneledDigitalInput.java:15]:
```suggestion
	public ChanneledDigitalInput(int channel, double debounceTime, boolean `returnFalseAsDefault`) {
```

maybe a bit overkill but I think it'll clarify some stuff.

**Review by Sh-Pe — CHANGES_REQUESTED** [review]

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+		this(booleanSupplier, debounceTime, debounceType, false);
+	}
+
+	@Override
+	public void updateInputs(DigitalInputInputsAutoLogged inputs) {
+		inputs.nonDebouncedValue = (inverted) ? !booleanSupplier.getAsBoolean() : booleanSupplier.getAsBoolean();
+		inputs.debouncedValue = debouncer.calculate(inputs.nonDebouncedValue);
+	}
+
+
```
Comment (greenblitz4590) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:37]: lmao

**Review by drorta — APPROVED** [review]
gg

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import edu.wpi.first.math.filter.Debouncer;
+import frc.utils.digitalinput.DigitalInputInputsAutoLogged;
+import frc.utils.digitalinput.IDigitalInput;
+
+import java.util.function.BooleanSupplier;
+
+public class SuppliedDigitalInput implements IDigitalInput {
+
+	private final Debouncer debouncer;
+	private boolean inverted;
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:12]: final

**Review by drorta — APPROVED** [review]
Gg nitay and maya

File: src/main/java/frc/utils/digitalinput/channeled/ChanneledDigitalInput.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+		this.inverted = inverted;
+	}
+
+	public ChanneledDigitalInput(int channel, double debounceTime) {
+		this(channel, debounceTime, false);
+	}
+
+	@Override
+	public void updateInputs(DigitalInputInputsAutoLogged inputs) {
+		inputs.nonDebouncedValue = (inverted) ? !digitalInput.get() : digitalInput.get();
```
Comment (bossyshrimp68) [inline src/main/java/frc/utils/digitalinput/channeled/ChanneledDigitalInput.java:26]: i dont think () is a must

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+		this.inverted = inverted;
+	}
+
+	public SuppliedDigitalInput(BooleanSupplier booleanSupplier, double debounceTime, Debouncer.DebounceType debounceType) {
+		this(booleanSupplier, debounceTime, debounceType, false);
+	}
+
+	@Override
+	public void updateInputs(DigitalInputInputsAutoLogged inputs) {
+		inputs.nonDebouncedValue = (inverted) ? !booleanSupplier.getAsBoolean() : booleanSupplier.getAsBoolean();
```
Comment (bossyshrimp68) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:33]: same here with ()

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Review by drorta — DISMISSED** [review]
gg

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+import frc.utils.digitalinput.IDigitalInput;
+
+import java.util.function.BooleanSupplier;
+
+public class SuppliedDigitalInput implements IDigitalInput {
+
+	private final Debouncer debouncer;
+	private final boolean inverted;
+
+	private final BooleanSupplier booleanSupplier;
```
Comment (hillelv) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:14]:
```suggestion
	private final BooleanSupplier booleanSupplier;
	private final Debouncer debouncer;
	private final boolean inverted;
```

**Review by bossyshrimp68 — DISMISSED** [review]

**Review by hillelv — DISMISSED** [review]

**Review by drorta — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

File: src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.utils.digitalinput.DigitalInputInputsAutoLogged;
+import frc.utils.digitalinput.IDigitalInput;
+
+public class ChooserDigitalInput implements IDigitalInput {
+
+	private final SendableChooser<Boolean> digitalInputValue;
+
+	public ChooserDigitalInput(String chooserName) {
+		this.digitalInputValue = new SendableChooser<>();
```
Comment (YHerm) [inline src/main/java/frc/utils/digitalinput/chooser/ChooserDigitalInput.java:13]: add line space after

File: src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+	private final Debouncer debouncer;
+	private final boolean inverted;
+	private final BooleanSupplier booleanSupplier;
+
+	public SuppliedDigitalInput(
+		BooleanSupplier booleanSupplier,
+		double debounceTime,
+		Debouncer.DebounceType debounceType,
+		boolean inverted
+	) {
```
Comment (bossyshrimp68) [inline src/main/java/frc/utils/digitalinput/supplied/SuppliedDigitalInput.java:20]: this should follow the order of the members of the class

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Review by drorta — APPROVED** [review]
Gg^3

**Pull request merged**